reset line distance when it is near max representable distance

ansis opened the pull request:
fix #2099 

When the distance of the current vertex is > half the maximum distance that can be represented, reset the distance. This fixes the original problem where a line with many segments is longer than the maximum distance.

We still need to make sure we can represent the distance of the longest possible segment so we need to make two more changes:
- improve how `linesofar` is packed into the buffer so that we have two more bits
- scale the distance down before adding it to the buffer to increase the maximum distance that can be represented

I don't like how `var distance` changed to `this.distance` but `addCurrentVertex` needs to be able to reset the distance and I'm not seeing a better way to do this.

:eyes: @lucaswoj

lucaswoj commented on js/data/line_bucket.js:
```diff
@@ -63,7 +68,7 @@ LineBucket.prototype.shaders = {
                 // so we need to shift the linesofar.
                 '((dir < 0) ? -1 : 1) * ((dir ? 1 : 0) | ((linesofar << 1) & 0x7F))',
-                '(linesofar >> 6) & 0x7F'
+                '((linesofar >> 6) & 0xFF) - 128'
```
Is this complex bit packing really necessary? If so, can you add a comment to the source about why it is necessary and how it works.

lucaswoj commented on js/data/line_bucket.js:
```diff
@@ -27,6 +27,11 @@ var EXTRUDE_SCALE = 63;
 
+// The maximum distance along a line that can be stored in the buffers.
+var MAX_LINE_DISTANCE = Math.pow(2, 14);
+var LINE_DISTANCE_SCALE = 1 / 2;
```
Please add a comment clarifying the unit of this value (even if it is "extent units")

lucaswoj commented on shaders/linepattern.vertex.glsl:
```diff
@@ -8,6 +8,10 @@ precision mediump float;
 #define scale 0.015873016
 
+// We scale the distance before adding it to the buffers so that we can store
+// long distances for long segments. Use this value to unscale the distance.
```
This comment would be useful in `LineBucket` too

lucaswoj commented on js/data/line_bucket.js:
```diff
@@ -175,8 +181,8 @@ LineBucket.prototype.addLine = function(vertices, join, cap, miterLimit, roundLi
                 var newPrevVertex = currentVertex.sub(currentVertex.sub(prevVertex)._mult(sharpCornerOffset / prevSegmentLength)._round());
-                distance += newPrevVertex.dist(prevVertex);
-                this.addCurrentVertex(newPrevVertex, flip, distance, prevNormal.mult(1), 0, 0, false);
+                this.distance += newPrevVertex.dist(prevVertex) * LINE_DISTANCE_SCALE;
```
Do we have to scale the distance here? Or can that be done [here](https://github.com/mapbox/mapbox-gl-js/pull/2114/files#diff-7ffec697472833439424ba5814cb0c00R71)? My intent was for all buffer packing craziness to go there.

lucaswoj commented on js/data/line_bucket.js:
```diff
@@ -377,6 +383,14 @@ LineBucket.prototype.addCurrentVertex = function(currentVertex, flip, distance,
+
+    // There is a maximum "distance along the line" that we can store in the buffers.
+    // When we get close to the distance, reset it to zero and add the vertex again with
+    // a distance of zero.
```
Please clarify that this maximum distance is determined by the number of bits in the `linesofar` attribute

lucaswoj commented:
> I don't like how `var distance` changed to `this.distance` but `addCurrentVertex` needs to be able to reset the distance and I'm not seeing a better way to do this.

I can live with this for now.

(Though I acutely feel the pain of this class's overcomplicated control flow and too-long methods)

ansis commented:
I added comments, moved LINE_DISTANCE_SCALE and added a couple more simplifications: 3b5bc9e 75ccf64

I closed this pr and opened a new one against master in https://github.com/mapbox/mapbox-gl-js/pull/2159
